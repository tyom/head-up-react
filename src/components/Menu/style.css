.menu {
  width: 180px;
  padding: 1rem 1rem 1rem 2.5rem;
  transition: ease-in-out 160ms;
  background-color: color(#000 a(.4));
}

.list {
  list-style: none;
  transition: 160ms;
  padding: 0;
  margin: 0;
}

.list > li {
  height: 10vh;
  border: 1px solid color(#fff a(.2));
  background-color: color(#fff a(.02));
  align-self: center;
  padding: .5rem;
  cursor: pointer;
  opacity: .7;
  transition: .2s;

  &:not(:first-child) {
    margin-top: 1rem;
  }

  &:hover {
    border: 1px solid color(#fff a(.2));
    background-color: color(#fff a(.05));
    box-shadow: 0 0 0 2px #17306e;
  }
}

.toggle {
  border-width: 0 1px 0 0;
  border-right: 1px solid color(#fff a(.1));
  color: #fff;
  background: none;
  position: absolute;
  z-index: 1;
  cursor: pointer;
  padding: 0;
  width: 1.5rem;
  top: 0;
  bottom: 0;
  left: 0;
  outline: none;
  transition: 200ms background-color;

  &:active {
    transition: none;
    background-color: #142a59;
  }

  > span {
    display: hidden;
  }
}

.toggleIcon {
  size: 1.3rem;
  opacity: .7;
  display: block;
  transition: 200ms;
}

.toggle:focus,
.toggle:hover {
  .toggleIcon {
    size: 1.8rem;
    margin-left: -0.3rem;
  }
}

/* States */

.menu.is-hidden {
  margin-left: calc(-180px + 1.5rem);

  .list {
    opacity: 0;
  }
  .toggle {
    border-color: transparent;
  }
}

.list li.is-active {
  border-color: #fff;
  opacity: 1;
  box-shadow: none;
}
