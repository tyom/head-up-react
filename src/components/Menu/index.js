import React from 'react';
import PropTypes from 'prop-types';
import kebabCase from 'lodash/kebabCase';
import classNames from 'classnames';

import Cell from '../Cell';
import Toggle from './Toggle';
import './style.css';

export default function Menu({
  items = [],
  isMenuClosed = false,
  activeDashboard = items[0],
  onSelectMenuItem,
  onToggleMenu,
}) {
  if (!items.length) {
    return;
  }

  return (
    <nav styleName={classNames('menu', { 'is-hidden': isMenuClosed })}>
      {onToggleMenu ? (
        <Toggle onClick={onToggleMenu} isMenuClosed={isMenuClosed} />
      ) : null}
      <ul styleName="list">
        {items.map(item => (
          <li
            key={kebabCase(item.name)}
            styleName={classNames('list-item', {
              'is-active': item.name === activeDashboard,
            })}
            onClick={() => onSelectMenuItem(item.name)}
          >
            {item.cells.map((cellSize, i) => (
              <Cell styleName="list-item-cell" size={cellSize} key={i} />
            ))}
          </li>
        ))}
      </ul>
    </nav>
  );
}

Menu.propTypes = {
  items: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      cells: PropTypes.array,
    })
  ),
  onSelectMenuItem: PropTypes.func,
  onToggleMenu: PropTypes.func,
  activeDashboard: PropTypes.string,
  isMenuClosed: PropTypes.bool,
};
