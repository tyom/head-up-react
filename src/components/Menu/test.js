import React from 'react';
import Menu from './';

const menuItems = [
  {
    name: 'first',
    cells: ['1', '2:1', '1:3'],
  },
  {
    name: 'second',
    cells: ['2', '3:2', '1:2'],
  },
];

const onToggleMenuSpy = jest.fn();
const onSelectMenuItemSpy = jest.fn();

const getItemsClassNames = arrayOfComponents =>
  arrayOfComponents.map(x => x.getElement().props.className);

describe('Menu: structure', () => {
  test("doesn't render if no items", () => {
    const component = shallow(<Menu />);
    expect(component).toMatchSnapshot();
  });

  test('renders menu items (dashboards) with cells', () => {
    const component = shallow(<Menu items={menuItems} />);
    expect(component).toMatchSnapshot();
  });

  test('renders menu as closed', () => {
    const component = shallow(<Menu items={menuItems} isMenuClosed={true} />);
    const el = component.getElement();
    expect(el.props.className).toMatchSnapshot();
  });

  test('active state', () => {
    const component1 = shallow(
      <Menu items={menuItems} activeDashboard="first" />
    );
    const component2 = shallow(
      <Menu items={menuItems} activeDashboard="second" />
    );
    const firstClassNames = getItemsClassNames(component1.find('ul > li'));
    const secondClassNames = getItemsClassNames(component2.find('ul > li'));

    expect(firstClassNames).toMatchSnapshot();
    expect(secondClassNames).toMatchSnapshot();
  });

  test('render toggleMenu open', () => {
    const component = shallow(
      <Menu items={menuItems} onToggleMenu={onToggleMenuSpy} />
    );

    expect(component.find('Toggle')).toMatchSnapshot();
  });

  test('render toggleMenu closed', () => {
    const component = shallow(
      <Menu
        items={menuItems}
        onToggleMenu={onToggleMenuSpy}
        isMenuClosed={true}
      />
    );

    expect(component.find('Toggle')).toMatchSnapshot();
  });
});

describe('Menu: behaviour', () => {
  test('menu toggle responds to clicks', () => {
    const component = shallow(
      <Menu items={menuItems} onToggleMenu={onToggleMenuSpy} />
    );

    component.find('Toggle').simulate('click');

    expect(onToggleMenuSpy).toBeCalled();
  });

  test('menu items respond to clicks', () => {
    const component = shallow(
      <Menu items={menuItems} onSelectMenuItem={onSelectMenuItemSpy} />
    );

    component
      .find('li')
      .at(0)
      .simulate('click');
    expect(onSelectMenuItemSpy).toBeCalledWith('first');

    component
      .find('li')
      .at(1)
      .simulate('click');
    expect(onSelectMenuItemSpy).toBeCalledWith('second');
  });
});
