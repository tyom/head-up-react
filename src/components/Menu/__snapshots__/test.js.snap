// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu: structure active state 1`] = `
Array [
  "style_list-item__1_OAQ style_is-active__3siLF",
  "style_list-item__1_OAQ",
]
`;

exports[`Menu: structure active state 2`] = `
Array [
  "style_list-item__1_OAQ",
  "style_list-item__1_OAQ style_is-active__3siLF",
]
`;

exports[`Menu: structure doesn't render if no items 1`] = `""`;

exports[`Menu: structure render toggleMenu closed 1`] = `
<Toggle
  isMenuClosed={true}
  onClick={[MockFunction]}
/>
`;

exports[`Menu: structure render toggleMenu open 1`] = `
<Toggle
  isMenuClosed={false}
  onClick={[MockFunction]}
/>
`;

exports[`Menu: structure renders menu as closed 1`] = `"style_menu__10DPV style_is-hidden__1nmAb"`;

exports[`Menu: structure renders menu items (dashboards) with cells 1`] = `
<nav
  className="style_menu__10DPV"
>
  <ul
    className="style_list__342LE"
  >
    <li
      className="style_list-item__1_OAQ"
      key="first"
      onClick={[Function]}
    >
      <Cell
        className="style_list-item-cell__W6X3Q"
        key="0"
        size="1"
      />
      <Cell
        className="style_list-item-cell__W6X3Q"
        key="1"
        size="2:1"
      />
      <Cell
        className="style_list-item-cell__W6X3Q"
        key="2"
        size="1:3"
      />
    </li>
    <li
      className="style_list-item__1_OAQ"
      key="second"
      onClick={[Function]}
    >
      <Cell
        className="style_list-item-cell__W6X3Q"
        key="0"
        size="2"
      />
      <Cell
        className="style_list-item-cell__W6X3Q"
        key="1"
        size="3:2"
      />
      <Cell
        className="style_list-item-cell__W6X3Q"
        key="2"
        size="1:2"
      />
    </li>
  </ul>
</nav>
`;
